rfcs/0000-windows-containers.md

```diff
@@ -0,0 +1,73 @@
+## Problem:
+
+kpack does not support building containers that can run on windows
+
+## Outcomes:
+
+### kpack should run on clusters that have windows workers
+
+kpack service pods and build pods for linux images will need to know not to run on the windows workers to work in a mixed-os cluster.
+
+The kpack release configuration should specify to run pods explicitly on linux os.
+The kpack controller should schedule linux builds on linux workers.
+
+**Actions to Take**
+
+Use the following node selector for all service pods and build pods for linux images:
+
+```
+nodeSelector:
+    kubernetes.io/os: linux
+```
+
+### kpack should support creating windows builders
+
+This is currently supported in the [pack cli](https://github.com/buildpacks/pack/issues/469)
+
+**Actions to Take**
+
+- kpack-controller will use a new environment variable `WINDOWS_LIFECYCLE_IMAGE` to reference the windows lifecycle image
+- When a Builder or ClusterBuilder is created using a windows stack and windows buildpackages, kpack-controller must create a windows builder using the windows lifecycle image.
+    - Image os is determined by the image [os property](https://github.com/opencontainers/image-spec/blob/master/config.md#:~:text=os%20string).
+
+Notes:
+
+- Builders with mixed os images must error gracefully
+- Builders should be labeled with `os.kpack.io: windows` for os disambiguation independent of image metadata
+
+### kpack should support running builds on windows workers
+
+**Actions to Take**
+
+- kpack-controller will accept new environment variables for windows images needed to run builds. These will have the same functionality as on linux and will be used for running builds on windows workers.
+    - `WINDOWS_BUILD_INIT_IMAGE` - a windows build init image
+    - `WINDOWS_COMPLETION_IMAGE` - a windows completion image
+    - `WINDOWS_REBASE_IMAGE` - a windows rebase image
+    - These will be optional to remove the dependency on windows images
```

Comment:
Is there a reason for these to be optional or could we always ship with them in our release.yaml? They will never get used/pulled if windows is not used.

Reply:
I don't see the need for them.

And I would be concerned with distribution rights maybe?

Reply:
@micahyoung Can correct me if I am wrong but, we wouldn't actually ship the windows layers (as they will be foreign layers) so, I don't there are distribution concerns.

Reply:
@matthewmcnew Well, all non-distributable baselayer blobs would live on `mcr.microsoft.com` and be referenced as a `url` in the image manifest layer, but any layers you create yourself above that would need to live as blobs in your registry alongside your the manifest. 

Here's the bottom two layers from `golang:1.15-nanoserver-1809` manifest for example: `layers[0]`'s blob would come from `mcr.microsoft.com` but `layers[1]`'s blob would come from the manifest'ss registry (`index.docker.io` in this case):
```
  "layers": [
    {
      "mediaType": "application/vnd.docker.image.rootfs.foreign.diff.tar.gzip",
      "size": 101205155,
      "digest": "sha256:aab6118ce69c93410df7fa15842a6e3b3c7ff20b639c779b5d5f78e7653eaa07",
      "urls": [
        "https://mcr.microsoft.com/v2/windows/nanoserver/blobs/sha256:aab6118ce69c93410df7fa15842a6e3b3c7ff20b639c779b5d5f78e7653eaa07"
      ]
    },
    {
      "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
      "size": 911,
      "digest": "sha256:b5bc5109798f9a64116d8a878ce7b6065052706b8c0b20c2a60b08b6e159c98e"
    },
    ...
  ]
```

Reply:
I think think that would prevent distribution concerns of publishing windows images as part of the kpack release.